### DISCUSSION TITLE
SageTV vs. Java

### DISCUSSION THREAD

**Discussion by GitOutWest** [OpenSageTV/Community#37] [Windows]
A couple days ago I let a Win10 Pro 22H2 system update Java from 7=>8 (really 1.7 to 1.8?). I didn't try SageTV x64 9.2.6 Client on that PC until yesterday when I found it wouldn't launch, throwing up a dialog complaining about needing Java greater-than-or-equal to 1.7. I tried a number of things, including deinstalling & reinstalling the SageTV x64 9.2.6 Client and uninstalling the latest 8/1.8 Java. Got messages about bad paths and missing jvm.dll when trying to run SageTV Cient from the LAN. I eventually got it working again after downloading Java 7/1.7 (x64 & x32 versions) directly from Oracle, installing those, *and* reinstalling Java 8/1.8 (x64 and x32 versions)--a shotgun solution. So both versions are now on this PC.

I had a similar Java problem with another Win11 x64 Pro system (where I installed the main SageTV x64 9.2.6 a couple months ago, replacing an old Win7 x64 Pro SageTV x64 "server" PC) a month ago, with the SageTVSetupx64_9.2.6.exe GUI refusing to run even though it had been running the GUI OK for a while. The Service, doing the actual recording etc, was running OK. I did some sort of Java reinstall & reboot (didn't note the steps carefully, sorry) and was again able to run the Sage GUI again.

So I'm just reporting that Sage is being "temperamental" about Java. This was something I hadn't run into on the previous Win7 server PC since it was so old it was no longer updating. I've been running a mix of Win7 Pro (now also replaced with a Win11 Pro PC) & Win10 Pro for the Client PCs and hadn't hit the Java problem previously. I'll have to be more aware and careful about not allowing Java updates.

**Comment by LanceUMatthews** [comment]
SageTV [looks for `jvm.dll`](https://github.com/google/sagetv/blob/bd0410e67ea5fed38dc828ba8d450e0eb9090267/native/exe/SageLauncher/launcher.cpp#L898) in one of two places:

1. `jre\bin\client\jvm.dll` in the application directory
2. Otherwise, the path contained in the `HKEY_LOCAL_MACHINE\SOFTWARE\JavaSoft\Java Runtime Environment\[VERSION]\RuntimeLib` registry value, where `[VERSION]` is the value at `HKEY_LOCAL_MACHINE\SOFTWARE\JavaSoft\Java Runtime Environment\CurrentVersion`.

   Note that if you are running 32-bit SageTV and, therefore, 32-bit Java on 64-bit Windows, the base key will be at `HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\JavaSoft\Java Runtime Environment\`.

If you get the "Java Missing" startup error, check that those two registry values exist and are correct.  Since I've started using the [Eclipse Temurin builds of Java 8/1.8](https://adoptium.net/temurin/releases/?os=windows&package=jre&version=8) I find that I still need to manually update the `CurrentVersion` value after installing minor Java upgrades, even despite enabling the JavaSoft registry keys feature in the installer.

**Comment by b-morgan** [comment]
What versions of Java does SageTV work with? I am currently running Java 8 Update 371 because a Java Update beyond that caused a problem. I'm not sure if it was just one particular update version or if it is anything beyond 371.
 
I now have an application that requires Java 17 (and runs on Java 22) so I'd like to know if I can configure SageTV to use a specific version of Java (and if so, how to configure it).

**Comment by GitOutWest** [comment]
I can't tell you the latest Java that works with Sage.

But I'm running my Win11 Sage server with Java 1.8.0_191, a Win10 Sage Client with Java 1.8.0_410 [from Sage TV's System Information pages on each] and another Sage Client on a Win11 PC with Java 1.8.0_191 [per Sage Sys Info page], although Sage reports that as Win10.

You can read above my original Java-hassle with Sage. Since then I've been trying to decline Java updates on my PCs so as to not break Sage.

**Comment by matthelm007** [comment]
BTW, in my case, the 64-bit Windows key is C:\Program Files\Java\jre-1.8\bin\server\jvm.dll and it use to be C:\Program Files\Java\jre-1.8\bin\client\jvm.dll   (Notice the change from client to server.)
